File: alb.py
#!/usr/bin/env python3

import boto3
from dotenv import load_dotenv
import os
import sys

class ALBUtils(object):
    def __init__(self):
        self.alb_client = None
        self.verbose = False

    def _create_client(self, region=None):
        self.alb_client = boto3.client('elbv2', region_name=region)

    def dns_name(self, vars_file=None, region=None, project=None, component=None, org=None):
        if vars_file is not None:
            load_dotenv(vars_file)
        region = region or os.environ.get('AWS_REGION')
        project = project or os.environ.get('PROJECT_NAME')
        luther_env = self._luther_env()
        self._create_client(region=region)
        match = {'Project': project, 'Environment': luther_env, 'Component': component, 'Organization': org}

        found = 0
        marker = None
        while True:
            args = {'PageSize': 20}
            if marker is not None:
                args['Marker'] = marker
            response = self.alb_client.describe_load_balancers(**args)
            albs = response.get('LoadBalancers', [])
            found += self._print_matching_albs(albs, match)
            marker = response.get('NextMarker')
            if marker is None:
                break
        if found == 0:
            sys.exit(1)

    def _print_matching_albs(self, albs, match):
        arns = [alb.get('LoadBalancerArn') for alb in albs]
        names = {alb.get('LoadBalancerArn'): alb.get('DNSName') for alb in albs}
        alb_tags = self.alb_client.describe_tags(ResourceArns=arns)
        found = 0
        for entry in alb_tags.get('TagDescriptions', []):
            tags = { item.get('Key'): item.get('Value') for item in entry.get('Tags', []) }
            if self.verbose:
                sys.stderr.write('{}\n'.format(tags))
            if self._matches_tags(tags, match):
                found += 1
                print(names[entry.get('ResourceArn')])
        return found

    def _matches_tags(self, tags, match):
        for k, v in match.items():
            if not self._matches_tag_value(tags, k, v):
                return False
        return True

    def _matches_tag_value(self, tags, tag_name, value):
        if value is None:
            return True
        return tags.get(tag_name) == value


    def _luther_env(self):
        if self.env == 'integration':
            return 'integ'
        return self.env


    def main(self):
        import argparse
        commonparser = argparse.ArgumentParser(add_help=False)
        commonparser.add_argument('--verbose', '-v', dest='verbosity', action='count', default=0)
        commonparser.add_argument('--region', help="aws region containing albs")
        argparser = argparse.ArgumentParser()
        argparser.add_argument('env', help='The project environment to use')
        subparsers = argparser.add_subparsers()
        dns_parser = subparsers.add_parser('alb-dns', parents=[commonparser])
        dns_parser.add_argument('--project', help='Only print albs with a matching project tag')
        dns_parser.add_argument('--component', help='Only print albs with a matching component tag')
        dns_parser.add_argument('--org', help='Only print albs with a matching organization tag')
        dns_parser.add_argument('--vars-file', '-f', help='File/Script with environment variable definitions')
        dns_parser.set_defaults(parser_func=self.dns_name)
        args = argparser.parse_args()

        self.env = args.env
        if args.verbosity > 0:
            self.verbose = True

        args_ignore = set(['parser_func', 'env', 'verbosity'])
        kwargs = {k: v for k, v in vars(args).items() if k not in args_ignore}

        args.parser_func(**kwargs)


if __name__ == '__main__':
    ALBUtils().main()

File: internal/alb/alb.go
package alb

import (
	"context"
	"fmt"
	"io"
	"os"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	elbv2 "github.com/aws/aws-sdk-go-v2/service/elasticloadbalancingv2"
	elbv2types "github.com/aws/aws-sdk-go-v2/service/elasticloadbalancingv2/types"
	"github.com/joho/godotenv"
	"github.com/luthersystems/mars/internal/app"
	"github.com/luthersystems/mars/internal/runner"
)

type DNSCmd struct {
	Verbose   int    `name:"verbose" short:"v" type:"counter" help:"Print matching diagnostics."`
	Region    string `name:"region" help:"AWS region containing ALBs."`
	Project   string `name:"project" help:"Only print ALBs with a matching project tag."`
	Component string `name:"component" help:"Only print ALBs with a matching component tag."`
	Org       string `name:"org" help:"Only print ALBs with a matching organization tag."`
	VarsFile  string `name:"vars-file" short:"f" help:"File with environment variable definitions."`
}

func (c *DNSCmd) Run(ctx context.Context, rt *app.Runtime) error {
	stdout := rt.Stdout
	if stdout == nil {
		stdout = os.Stdout
	}
	stderr := rt.Stderr
	if stderr == nil {
		stderr = os.Stderr
	}
	if c.VarsFile != "" {
		if err := godotenv.Load(c.VarsFile); err != nil {
			return err
		}
	}
	region := c.Region
	if region == "" {
		region = os.Getenv("AWS_REGION")
	}
	project := c.Project
	if project == "" {
		project = os.Getenv("PROJECT_NAME")
	}
	cfg, err := config.LoadDefaultConfig(ctx, config.WithRegion(region))
	if err != nil {
		return err
	}
	client := elbv2.NewFromConfig(cfg)
	match := map[string]string{
		"Project":      project,
		"Environment":  lutherEnv(rt.Target),
		"Component":    c.Component,
		"Organization": c.Org,
	}
	found, err := printMatching(ctx, client, stdout, stderr, match, c.Verbose > 0)
	if err != nil {
		return err
	}
	if found == 0 {
		return runner.Exit(1)
	}
	return nil
}

type elbClient interface {
	DescribeLoadBalancers(context.Context, *elbv2.DescribeLoadBalancersInput, ...func(*elbv2.Options)) (*elbv2.DescribeLoadBalancersOutput, error)
	DescribeTags(context.Context, *elbv2.DescribeTagsInput, ...func(*elbv2.Options)) (*elbv2.DescribeTagsOutput, error)
}

func printMatching(ctx context.Context, client elbClient, stdout io.Writer, stderr io.Writer, match map[string]string, verbose bool) (int, error) {
	found := 0
	var marker *string
	for {
		resp, err := client.DescribeLoadBalancers(ctx, &elbv2.DescribeLoadBalancersInput{
			Marker:   marker,
			PageSize: aws.Int32(20),
		})
		if err != nil {
			return 0, err
		}
		pageFound, err := printMatchingPage(ctx, client, stdout, stderr, resp.LoadBalancers, match, verbose)
		if err != nil {
			return 0, err
		}
		found += pageFound
		marker = resp.NextMarker
		if marker == nil {
			break
		}
	}
	return found, nil
}

func printMatchingPage(ctx context.Context, client elbClient, stdout io.Writer, stderr io.Writer, albs []elbv2types.LoadBalancer, match map[string]string, verbose bool) (int, error) {
	arns := make([]string, 0, len(albs))
	names := make(map[string]string, len(albs))
	for _, lb := range albs {
		arn := aws.ToString(lb.LoadBalancerArn)
		arns = append(arns, arn)
		names[arn] = aws.ToString(lb.DNSName)
	}
	if len(arns) == 0 {
		return 0, nil
	}
	resp, err := client.DescribeTags(ctx, &elbv2.DescribeTagsInput{ResourceArns: arns})
	if err != nil {
		return 0, err
	}
	found := 0
	for _, entry := range resp.TagDescriptions {
		tags := map[string]string{}
		for _, tag := range entry.Tags {
			tags[aws.ToString(tag.Key)] = aws.ToString(tag.Value)
		}
		if verbose {
			fmt.Fprintln(stderr, tags)
		}
		if matchesTags(tags, match) {
			found++
			fmt.Fprintln(stdout, names[aws.ToString(entry.ResourceArn)])
		}
	}
	return found, nil
}

func matchesTags(tags map[string]string, match map[string]string) bool {
	for key, value := range match {
		if value == "" {
			continue
		}
		if tags[key] != value {
			return false
		}
	}
	return true
}

func lutherEnv(env string) string {
	if env == "integration" {
		return "integ"
	}
	return env
}
